repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    @Override
+    public Optional<Certificate> findById(long id) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource().addValue(ID_PARAMETER, id);
+        List<Certificate> certificates = namedParameterJdbcTemplate.query(SELECT_CERTIFICATE_BY_ID, parameterSource, rowMapper);
+
+        return Optional.ofNullable(certificates.size() == 1 ? certificates.get(0) : null);
+    }
+
+    @Override
+    public boolean attachTag(long certificateId, long tagId) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(GIFT_CERTIFICATE_ID_PARAMETER, certificateId)
+                .addValue(TAG_ID_PARAMETER, tagId);
+
+        return namedParameterJdbcTemplate.update(INSERT_TAG_TO_CERTIFICATE, parameterSource) > 0;
+    }
+
+    @Override
+    public boolean detachTag(long certificateId, long tagId) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(GIFT_CERTIFICATE_ID_PARAMETER, certificateId)
+                .addValue(TAG_ID_PARAMETER, tagId);
+
+        return namedParameterJdbcTemplate.update(DELETE_TAG_FROM_CERTIFICATE, parameterSource) > 0;
+    }
+
+    @Override
+    public void create(Certificate certificate) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(NAME_PARAMETER, certificate.getName())
+                .addValue(DESCRIPTION_PARAMETER, certificate.getDescription())
+                .addValue(PRICE_PARAMETER, certificate.getPrice())
+                .addValue(DURATION_PARAMETER, certificate.getDuration().toDays())
+                .addValue(CREATE_DATE_PARAMETER, certificate.getCreateDate())
+                .addValue(LAST_UPDATE_PARAMETER, certificate.getLastUpdateDate());
+
+        namedParameterJdbcTemplate.update(INSERT_CERTIFICATE, parameterSource);
+    }
+
+    @Override
+    public boolean update(Certificate certificate) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(NAME_PARAMETER, certificate.getName())
+                .addValue(DESCRIPTION_PARAMETER, certificate.getDescription())
+                .addValue(PRICE_PARAMETER, certificate.getPrice())
+                .addValue(DURATION_PARAMETER, certificate.getDuration().toDays())
+                .addValue(CREATE_DATE_PARAMETER, certificate.getCreateDate())
+                .addValue(LAST_UPDATE_PARAMETER, certificate.getLastUpdateDate());
+
+        return namedParameterJdbcTemplate.update(UPDATE_CERTIFICATE, parameterSource) > 0;
+    }
+
+    @Override
+    public boolean delete(long id) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource().addValue(ID_PARAMETER, id);
+
+        return namedParameterJdbcTemplate.update(DELETE_CERTIFICATE, parameterSource) > 0;
+    }
+
+    private String sqlBuild(List<String> where, List<String> order) {

```

Comment:
Let's think about putting the logic for building a query with parameters into a separate class. This is a separate whole functionality